// @colyseus/schema/lib/types/symbols.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.$fieldIndexesByViewTag = exports.$viewFieldIndexes = exports.$refTypeFieldIndexes = exports.$numFields = exports.$descriptors = exports.$onDecodeEnd = exports.$onEncodeEnd = exports.$childType = exports.$changes = exports.$deleteByIndex = exports.$getByIndex = exports.$filter = exports.$decoder = exports.$encoder = exports.$track = void 0;
exports.$track = "~track";
exports.$encoder = "~encoder";
exports.$decoder = "~decoder";
exports.$filter = "~filter";
exports.$getByIndex = "~getByIndex";
exports.$deleteByIndex = "~deleteByIndex";
/**
 * Used to hold ChangeTree instances whitin the structures
 */
exports.$changes = '~changes';
/**
 * Used to keep track of the type of the child elements of a collection
 * (MapSchema, ArraySchema, etc.)
 */
exports.$childType = '~childType';
/**
 * Optional "discard" method for custom types (ArraySchema)
 * (Discards changes for next serialization)
 */
exports.$onEncodeEnd = '~onEncodeEnd';
/**
 * When decoding, this method is called after the instance is fully decoded
 */
exports.$onDecodeEnd = "~onDecodeEnd";
/**
 * Metadata
 */
exports.$descriptors = "~descriptors";
exports.$numFields = "~__numFields";
exports.$refTypeFieldIndexes = "~__refTypeFieldIndexes";
exports.$viewFieldIndexes = "~__viewFieldIndexes";
exports.$fieldIndexesByViewTag = "$__fieldIndexesByViewTag";
