// @unboxy/phaser-sdk/dist/realtime/RealtimeModule.js
import { RpcError } from '../core/Transport.js';
import { UnboxyRoom } from './UnboxyRoom.js';
/**
 * Multiplayer rooms backed by unboxy-realtime-service (Colyseus under the hood).
 *
 * Availability: only when the host provided a `realtimeUrl` during the
 * handshake (i.e. connected to unboxy-home-ui with multiplayer enabled).
 * Unavailable in standalone/anonymous mode — methods throw RpcError
 * 'REALTIME_UNAVAILABLE'.
 *
 * Tokens are fetched per connection attempt via the `realtime.getToken`
 * RPC — the iframe never holds a long-lived credential.
 */
export class RealtimeModule {
    constructor(transport, realtimeUrl) {
        this.transport = transport;
        this.realtimeUrl = realtimeUrl;
        this.client = null;
    }
    /** `true` when a realtime endpoint was provided by the host. */
    get available() {
        return typeof this.realtimeUrl === 'string' && this.realtimeUrl.length > 0;
    }
    async joinOrCreate(roomType, options = {}) {
        return this.connect(roomType, options, 'joinOrCreate');
    }
    async create(roomType, options = {}) {
        return this.connect(roomType, options, 'create');
    }
    async join(roomType, options = {}) {
        return this.connect(roomType, options, 'join');
    }
    async joinById(roomId, options = {}) {
        return this.connect(roomId, options, 'joinById');
    }
    async connect(nameOrId, options, method) {
        if (!this.available) {
            throw new RpcError('REALTIME_UNAVAILABLE', 'Multiplayer is not available for this host. Unboxy.rooms requires running inside unboxy-home-ui with realtime enabled.');
        }
        if (!this.client) {
            // Dynamic import so single-player games do not pull Colyseus into their
            // bundle. Vite tree-shakes this path when `rooms` is never referenced.
            const { Client } = await import('colyseus.js');
            this.client = new Client(this.realtimeUrl);
        }
        const { token } = await this.transport.call('realtime.getToken', {});
        // Inject gameId from the transport so the server-side gameId check sees
        // a value consistent with the JWT — games never pass this directly.
        const merged = { ...options, gameId: this.transport.gameId };
        this.client.auth.token = token;
        const room = await this.client[method](nameOrId, merged);
        return new UnboxyRoom(room);
    }
}
