// @unboxy/phaser-sdk/dist/realtime/UnboxyRoom.js
/**
 * Unboxy-flavored wrapper around a Colyseus Room. Exposes only the surface
 * documented in SDK-GUIDE.md so games are portable to a different realtime
 * backend should we migrate away from Colyseus.
 */
export class UnboxyRoom {
    constructor(room) {
        this.room = room;
    }
    get id() { return this.room.roomId; }
    get name() { return this.room.name; }
    get sessionId() { return this.room.sessionId; }
    /**
     * Current server-authoritative state. Proxied from Colyseus Schema — read
     * values directly. Mutations do not propagate; only server-side handlers
     * may change state.
     */
    get state() { return this.room.state; }
    /** Send a typed message to the server. */
    send(type, payload) {
        this.room.send(type, payload);
    }
    /** Register a handler for a server-sent message type. Returns unsubscribe. */
    on(type, handler) {
        return this.room.onMessage(type, handler);
    }
    /** Fires whenever the server-authoritative state changes. */
    onStateChange(handler) {
        const cb = () => handler(this.room.state);
        this.room.onStateChange(cb);
        return () => this.room.onStateChange.remove(cb);
    }
    /**
     * Fires when the connection closes (kick, server shutdown, network drop).
     * `code` follows WebSocket close codes plus Colyseus-specific ones.
     */
    onLeave(handler) {
        const cb = (code) => handler(code);
        this.room.onLeave(cb);
        return () => this.room.onLeave.remove(cb);
    }
    /** Fires when the server reports an error for this room. */
    onError(handler) {
        const cb = (code, message) => handler(code, message);
        this.room.onError(cb);
        return () => this.room.onError.remove(cb);
    }
    /** Disconnect from the room. Resolves with the close code. */
    async leave(consented = true) {
        return this.room.leave(consented);
    }
}
