// @unboxy/phaser-sdk/dist/realtime/UnboxyRoom.js
function parseOrNull(raw) {
    if (raw == null)
        return null;
    try {
        return JSON.parse(raw);
    }
    catch {
        return null;
    }
}
/**
 * Facade for per-player state — one entry per player, only the owner can write
 * to their own entry. Delta-synced via Colyseus schema.
 */
export class PlayerDataFacade {
    constructor(room) {
        this.room = room;
    }
    /** Write a JSON-serializable value to the caller's own player data. */
    set(key, value) {
        this.room.send('player.set', { key, value });
    }
    /** Remove a key from the caller's own player data. */
    delete(key) {
        this.room.send('player.delete', { key });
    }
    /** Read a player's data value. Returns null if the player or key is absent. */
    get(sessionId, key) {
        const state = this.room.state;
        const player = state?.players?.get?.(sessionId);
        return parseOrNull(player?.data?.get?.(key));
    }
}
/**
 * Facade for room-scope shared state — one map per room, any client may write.
 * Delta-synced via Colyseus schema. Use for shared game state like current
 * turn, scoreboard, shared level-of-the-day, etc.
 */
export class RoomDataFacade {
    constructor(room) {
        this.room = room;
    }
    /** Write a JSON-serializable value into the room's shared data map. */
    set(key, value) {
        this.room.send('room.set', { key, value });
    }
    /** Remove a key from the room's shared data map. */
    delete(key) {
        this.room.send('room.delete', { key });
    }
    /** Read a shared room value. Returns null if the key is absent. */
    get(key) {
        const state = this.room.state;
        return parseOrNull(state?.data?.get?.(key));
    }
}
/**
 * Unboxy-flavored wrapper around a Colyseus Room. Exposes only the surface
 * documented in SDK-GUIDE.md so games are portable to a different realtime
 * backend should we migrate away from Colyseus.
 */
export class UnboxyRoom {
    constructor(room) {
        this.room = room;
        this.player = new PlayerDataFacade(room);
        this.data = new RoomDataFacade(room);
    }
    get id() { return this.room.roomId; }
    get name() { return this.room.name; }
    get sessionId() { return this.room.sessionId; }
    /**
     * Current server-authoritative state. Proxied from Colyseus Schema — read
     * values directly. Mutations do not propagate; only server-side handlers
     * may change state.
     */
    get state() { return this.room.state; }
    /** Send a transient message. Relayed to every other client in the room
     *  with `from: sessionId` stamped onto the payload. Not persisted in state. */
    send(type, payload) {
        this.room.send(type, payload);
    }
    /** Register a handler for a server-sent message type. Returns unsubscribe. */
    on(type, handler) {
        return this.room.onMessage(type, handler);
    }
    /** Fires whenever the server-authoritative state changes. */
    onStateChange(handler) {
        const cb = () => handler(this.room.state);
        this.room.onStateChange(cb);
        return () => this.room.onStateChange.remove(cb);
    }
    /**
     * Fires when the connection closes (kick, server shutdown, network drop).
     * `code` follows WebSocket close codes plus Colyseus-specific ones.
     */
    onLeave(handler) {
        const cb = (code) => handler(code);
        this.room.onLeave(cb);
        return () => this.room.onLeave.remove(cb);
    }
    /** Fires when the server reports an error for this room. */
    onError(handler) {
        const cb = (code, message) => handler(code, message);
        this.room.onError(cb);
        return () => this.room.onError.remove(cb);
    }
    /** Disconnect from the room. Resolves with the close code. */
    async leave(consented = true) {
        return this.room.leave(consented);
    }
}
