// @unboxy/phaser-sdk/dist/core/UnboxyScene.js
import Phaser from 'phaser';
/**
 * Base scene class with common Unboxy utilities.
 * Extend this instead of Phaser.Scene for built-in helpers.
 */
export class UnboxyScene extends Phaser.Scene {
    /**
     * Create a simple text button with hover effects.
     */
    createButton(x, y, text, style, onClick) {
        const btn = this.add.text(x, y, text, {
            fontSize: '20px',
            color: '#ffffff',
            backgroundColor: '#4662D8',
            padding: { x: 16, y: 8 },
            ...style,
        })
            .setOrigin(0.5)
            .setInteractive({ useHandCursor: true });
        btn.on('pointerover', () => btn.setAlpha(0.8));
        btn.on('pointerout', () => btn.setAlpha(1));
        if (onClick)
            btn.on('pointerdown', onClick);
        return btn;
    }
    /**
     * Shake the camera briefly (e.g., on damage).
     */
    shakeCamera(duration = 100, intensity = 0.01) {
        this.cameras.main.shake(duration, intensity);
    }
    /**
     * Flash the camera briefly (e.g., on hit).
     */
    flashCamera(duration = 100, r = 255, g = 255, b = 255) {
        this.cameras.main.flash(duration, r, g, b);
    }
}
