// @unboxy/phaser-sdk/dist/core/UnboxyGame.js
import Phaser from 'phaser';
import { setupScreenshotListener } from '../screenshot/ScreenshotManager.js';
/**
 * Create an Unboxy-enhanced Phaser game instance.
 * Includes built-in integrations: screenshot capture, preserveDrawingBuffer, etc.
 */
export function createUnboxyGame(options) {
    const config = {
        type: Phaser.AUTO,
        backgroundColor: options.backgroundColor ?? '#1a1a2e',
        scale: {
            mode: Phaser.Scale.FIT,
            autoCenter: Phaser.Scale.CENTER_BOTH,
            width: options.width,
            height: options.height,
        },
        render: {
            preserveDrawingBuffer: true,
        },
        pixelArt: options.pixelArt ?? false,
        physics: options.physics ?? {
            default: 'arcade',
            arcade: { gravity: { x: 0, y: 0 }, debug: false },
        },
        scene: options.scenes,
    };
    const game = new Phaser.Game(config);
    // Built-in integrations
    setupScreenshotListener(game);
    return game;
}
