// @unboxy/phaser-sdk/dist/core/Unboxy.js
import { PostMessageTransport } from './transports/PostMessageTransport.js';
import { LocalStorageTransport } from './transports/LocalStorageTransport.js';
import { SavesModule } from '../saves/SavesModule.js';
import { GameDataModule } from '../gamedata/GameDataModule.js';
import { RealtimeModule } from '../realtime/RealtimeModule.js';
// Kept in sync with package.json on each publish.
const SDK_VERSION = '0.2.8';
/**
 * Unboxy platform services bound to the current (game, user).
 *
 * The public API is host-agnostic. Under the hood, `Unboxy.init()` detects the
 * host (home-ui iframe, Discord Activity, or standalone) and picks the right
 * transport. Games should not care which one is active.
 */
export class Unboxy {
    constructor(transport) {
        this.transport = transport;
        // Saves are per-user. When the viewer is anonymous, route save calls to
        // localStorage even if a host is attached — the backend requires auth,
        // so forwarding would 401. Games see the same `unboxy.saves` API either
        // way and don't have to branch on auth state.
        const savesTransport = transport.user === null
            ? new LocalStorageTransport(transport.gameId || 'anonymous')
            : transport;
        this.saves = new SavesModule(savesTransport);
        // gameData is game-scope: reads are public (anonymous OK), only writes
        // need auth. The primary transport handles both; if it's a LocalStorage
        // fallback (standalone mode) that also works — the module speaks the
        // same RPC interface.
        this.gameData = new GameDataModule(transport);
        // Multiplayer requires a realtime endpoint from the host. When absent
        // (standalone or a host without multiplayer), methods throw a clear
        // REALTIME_UNAVAILABLE error rather than silently misbehaving.
        this.rooms = new RealtimeModule(transport, transport.realtimeUrl);
    }
    /** The current authenticated user, or `null` if anonymous / standalone. */
    get user() { return this.transport.user; }
    get isAuthenticated() { return this.transport.user !== null; }
    /** Stable game identifier issued by the host. May be a placeholder when standalone. */
    get gameId() { return this.transport.gameId; }
    /** Which transport Unboxy.init() selected. Useful for debugging. */
    get host() { return this.transport.kind; }
    /**
     * Initialize the SDK. Detects the host and negotiates identity.
     *
     * Resolution order:
     *   1. If embedded in a window that responds to the Unboxy handshake → PostMessageTransport
     *   2. Otherwise → LocalStorageTransport (anonymous, local-only saves)
     *
     * Discord Activity support is designed for but not shipped in this version.
     */
    static async init(options = {}) {
        const handshakeTimeoutMs = options.handshakeTimeoutMs ?? 5000;
        const standaloneGameId = options.standaloneGameId ?? 'standalone';
        const pm = await PostMessageTransport.connect(handshakeTimeoutMs, SDK_VERSION);
        if (pm)
            return new Unboxy(pm);
        return new Unboxy(new LocalStorageTransport(standaloneGameId));
    }
}
