// @unboxy/phaser-sdk/dist/core/transports/LocalStorageTransport.js
import { RpcError } from '../Transport.js';
/**
 * Fallback transport when the game runs without a host (e.g., opened directly
 * outside unboxy-home-ui) or when the viewer is anonymous. Data lives in
 * localStorage, keyed by gameId. No network calls. Quotas mirror the backend.
 */
export class LocalStorageTransport {
    constructor(gameId) {
        this.gameId = gameId;
        this.kind = 'standalone';
        this.user = null;
    }
    async call(method, params) {
        switch (method) {
            case 'saves.get': return this.kvGet(this.savesPrefix(), params);
            case 'saves.set': return this.kvSet(this.savesPrefix(), params);
            case 'saves.delete': return this.kvDelete(this.savesPrefix(), params);
            case 'saves.list': return this.kvList(this.savesPrefix());
            case 'gameData.get': return this.kvGet(this.gameDataPrefix(), params);
            case 'gameData.set': return this.kvSet(this.gameDataPrefix(), params);
            case 'gameData.delete': return this.kvDelete(this.gameDataPrefix(), params);
            case 'gameData.list': return this.kvList(this.gameDataPrefix());
            default: throw new RpcError('UNKNOWN_METHOD', `Unknown method: ${method}`);
        }
    }
    savesPrefix() { return `unboxy:saves:${this.gameId}:`; }
    gameDataPrefix() { return `unboxy:gameData:${this.gameId}:`; }
    read(prefix, key) {
        if (typeof localStorage === 'undefined')
            return null;
        const raw = localStorage.getItem(prefix + key);
        if (!raw)
            return null;
        try {
            return JSON.parse(raw);
        }
        catch {
            return null;
        }
    }
    write(prefix, key, rec) {
        if (typeof localStorage === 'undefined')
            throw new RpcError('UNAVAILABLE', 'localStorage unavailable');
        localStorage.setItem(prefix + key, JSON.stringify(rec));
    }
    kvGet(prefix, { key }) {
        const rec = this.read(prefix, key);
        return rec ? { value: rec.value, version: rec.version } : { value: null, version: null };
    }
    kvSet(prefix, { key, value, ifVersion }) {
        const existing = this.read(prefix, key);
        if (ifVersion !== undefined && existing && existing.version !== ifVersion) {
            throw new RpcError('VERSION_MISMATCH', `Expected version ${ifVersion} but stored is ${existing.version}`);
        }
        const next = { value, version: (existing?.version ?? 0) + 1 };
        this.write(prefix, key, next);
        return { version: next.version };
    }
    kvDelete(prefix, { key }) {
        if (typeof localStorage === 'undefined')
            return { deleted: false };
        const full = prefix + key;
        const existed = localStorage.getItem(full) !== null;
        localStorage.removeItem(full);
        return { deleted: existed };
    }
    kvList(prefix) {
        if (typeof localStorage === 'undefined')
            return { keys: [] };
        const keys = [];
        for (let i = 0; i < localStorage.length; i++) {
            const k = localStorage.key(i);
            if (k && k.startsWith(prefix))
                keys.push(k.slice(prefix.length));
        }
        return { keys };
    }
}
