// @unboxy/phaser-sdk/dist/core/transports/LocalStorageTransport.js
import { RpcError } from '../Transport.js';
/**
 * Fallback transport when the game runs without a host (e.g., opened directly
 * outside unboxy-home-ui) or when the viewer is anonymous. Data lives in
 * localStorage, keyed by gameId. No network calls. Quotas mirror the backend.
 */
export class LocalStorageTransport {
    constructor(gameId) {
        this.gameId = gameId;
        this.kind = 'standalone';
        this.user = null;
    }
    async call(method, params) {
        switch (method) {
            case 'saves.get': return this.savesGet(params);
            case 'saves.set': return this.savesSet(params);
            case 'saves.delete': return this.savesDelete(params);
            case 'saves.list': return this.savesList();
            default: throw new RpcError('UNKNOWN_METHOD', `Unknown method: ${method}`);
        }
    }
    prefix() { return `unboxy:saves:${this.gameId}:`; }
    read(key) {
        if (typeof localStorage === 'undefined')
            return null;
        const raw = localStorage.getItem(this.prefix() + key);
        if (!raw)
            return null;
        try {
            return JSON.parse(raw);
        }
        catch {
            return null;
        }
    }
    write(key, rec) {
        if (typeof localStorage === 'undefined')
            throw new RpcError('UNAVAILABLE', 'localStorage unavailable');
        localStorage.setItem(this.prefix() + key, JSON.stringify(rec));
    }
    savesGet({ key }) {
        const rec = this.read(key);
        return rec ? { value: rec.value, version: rec.version } : { value: null, version: null };
    }
    savesSet({ key, value, ifVersion }) {
        const existing = this.read(key);
        if (ifVersion !== undefined && existing && existing.version !== ifVersion) {
            throw new RpcError('VERSION_MISMATCH', `Expected version ${ifVersion} but stored is ${existing.version}`);
        }
        const next = { value, version: (existing?.version ?? 0) + 1 };
        this.write(key, next);
        return { version: next.version };
    }
    savesDelete({ key }) {
        if (typeof localStorage === 'undefined')
            return { deleted: false };
        const full = this.prefix() + key;
        const existed = localStorage.getItem(full) !== null;
        localStorage.removeItem(full);
        return { deleted: existed };
    }
    savesList() {
        if (typeof localStorage === 'undefined')
            return { keys: [] };
        const keys = [];
        const p = this.prefix();
        for (let i = 0; i < localStorage.length; i++) {
            const k = localStorage.key(i);
            if (k && k.startsWith(p))
                keys.push(k.slice(p.length));
        }
        return { keys };
    }
}
