// @unboxy/phaser-sdk/dist/core/transports/PostMessageTransport.js
import { RpcError } from '../Transport.js';
import { PROTOCOL_VERSION, } from '../../protocol.js';
/**
 * Connects to a parent window that implements the Unboxy RPC host protocol.
 * Used when the game is loaded inside unboxy-home-ui.
 */
export class PostMessageTransport {
    constructor(parent) {
        this.parent = parent;
        this.kind = 'unboxy-home-ui';
        this.user = null;
        this.gameId = '';
        this.pending = new Map();
        this.seq = 0;
    }
    /**
     * Perform the handshake. Resolves once parent has replied with `unboxy:init`.
     * Resolves to null if no response within `timeoutMs`.
     *
     * The hello is retried at `helloRetryMs` intervals because the parent's
     * RPC-host message listener may not be mounted yet when the iframe first
     * fires hello — especially on slower mobile devices where React takes
     * longer to boot. One-shot hello was losing the handshake on Android.
     */
    static async connect(timeoutMs = 5000, sdkVersion = '0.0.0', helloRetryMs = 200) {
        if (typeof window === 'undefined' || window.parent === window)
            return null;
        const transport = new PostMessageTransport(window.parent);
        const hello = {
            type: 'unboxy:hello',
            protocolVersion: PROTOCOL_VERSION,
            sdkVersion,
        };
        const init = await new Promise((resolve) => {
            let settled = false;
            const finish = (value) => {
                if (settled)
                    return;
                settled = true;
                window.removeEventListener('message', onMessage);
                clearInterval(helloInterval);
                clearTimeout(timeoutHandle);
                resolve(value);
            };
            const onMessage = (event) => {
                if (event.source !== window.parent)
                    return;
                const data = event.data;
                if (!data || data.type !== 'unboxy:init')
                    return;
                finish(data);
            };
            window.addEventListener('message', onMessage);
            // Fire the first hello immediately, then keep retrying until init
            // arrives or timeout fires. Retries are cheap (a postMessage with a
            // small payload); the parent's RPC host replies on first valid hello.
            transport.parent.postMessage(hello, '*');
            const helloInterval = setInterval(() => {
                transport.parent.postMessage(hello, '*');
            }, helloRetryMs);
            const timeoutHandle = setTimeout(() => finish(null), timeoutMs);
        });
        if (!init)
            return null;
        if (init.protocolVersion !== PROTOCOL_VERSION) {
            console.warn('[UnboxySDK] protocol version mismatch', init.protocolVersion, PROTOCOL_VERSION);
        }
        transport.user = init.user;
        transport.gameId = init.gameId;
        transport.realtimeUrl = init.realtimeUrl;
        transport.installResultListener();
        return transport;
    }
    installResultListener() {
        window.addEventListener('message', (event) => {
            if (event.source !== this.parent)
                return;
            const data = event.data;
            if (!data || data.type !== 'unboxy:rpc.result')
                return;
            const pending = this.pending.get(data.id);
            if (!pending)
                return;
            this.pending.delete(data.id);
            if (data.ok)
                pending.resolve(data.result);
            else
                pending.reject(new RpcError(data.error.code, data.error.message));
        });
    }
    call(method, params) {
        const id = `${++this.seq}-${Date.now()}`;
        const message = { type: 'unboxy:rpc', id, method, params };
        return new Promise((resolve, reject) => {
            this.pending.set(id, { resolve: resolve, reject });
            this.parent.postMessage(message, '*');
            setTimeout(() => {
                if (this.pending.has(id)) {
                    this.pending.delete(id);
                    reject(new RpcError('TIMEOUT', `RPC ${method} timed out`));
                }
            }, 10000);
        });
    }
}
