// @unboxy/phaser-sdk/dist/core/transports/PostMessageTransport.js
import { RpcError } from '../Transport.js';
import { PROTOCOL_VERSION, } from '../../protocol.js';
/**
 * Connects to a parent window that implements the Unboxy RPC host protocol.
 * Used when the game is loaded inside unboxy-home-ui.
 */
export class PostMessageTransport {
    constructor(parent) {
        this.parent = parent;
        this.kind = 'unboxy-home-ui';
        this.user = null;
        this.gameId = '';
        this.pending = new Map();
        this.seq = 0;
    }
    /**
     * Perform the handshake. Resolves once parent has replied with `unboxy:init`.
     * Rejects if no response within `timeoutMs`.
     */
    static async connect(timeoutMs = 2000, sdkVersion = '0.0.0') {
        if (typeof window === 'undefined' || window.parent === window)
            return null;
        const transport = new PostMessageTransport(window.parent);
        const initPromise = new Promise((resolve) => {
            const onMessage = (event) => {
                if (event.source !== window.parent)
                    return;
                const data = event.data;
                if (!data || data.type !== 'unboxy:init')
                    return;
                window.removeEventListener('message', onMessage);
                resolve(data);
            };
            window.addEventListener('message', onMessage);
            setTimeout(() => {
                window.removeEventListener('message', onMessage);
                resolve(null);
            }, timeoutMs);
        });
        const hello = {
            type: 'unboxy:hello',
            protocolVersion: PROTOCOL_VERSION,
            sdkVersion,
        };
        transport.parent.postMessage(hello, '*');
        const init = await initPromise;
        if (!init)
            return null;
        if (init.protocolVersion !== PROTOCOL_VERSION) {
            console.warn('[UnboxySDK] protocol version mismatch', init.protocolVersion, PROTOCOL_VERSION);
        }
        transport.user = init.user;
        transport.gameId = init.gameId;
        transport.realtimeUrl = init.realtimeUrl;
        transport.installResultListener();
        return transport;
    }
    installResultListener() {
        window.addEventListener('message', (event) => {
            if (event.source !== this.parent)
                return;
            const data = event.data;
            if (!data || data.type !== 'unboxy:rpc.result')
                return;
            const pending = this.pending.get(data.id);
            if (!pending)
                return;
            this.pending.delete(data.id);
            if (data.ok)
                pending.resolve(data.result);
            else
                pending.reject(new RpcError(data.error.code, data.error.message));
        });
    }
    call(method, params) {
        const id = `${++this.seq}-${Date.now()}`;
        const message = { type: 'unboxy:rpc', id, method, params };
        return new Promise((resolve, reject) => {
            this.pending.set(id, { resolve: resolve, reject });
            this.parent.postMessage(message, '*');
            setTimeout(() => {
                if (this.pending.has(id)) {
                    this.pending.delete(id);
                    reject(new RpcError('TIMEOUT', `RPC ${method} timed out`));
                }
            }, 10000);
        });
    }
}
